src/types/Typechecker/Typechecker.hs

```diff
@@ -1722,33 +1722,43 @@ instance Checkable Expr where
                                         ,stop = eStop
                                         ,step = eStep}
 
-    --  E |- rng : Range
-    --  E, x : int |- e : ty
-    -- --------------------------
-    --  E |- for x <- rng e : ty
-
-    --  E |- arr : [ty]
-    --  E, x : int |- e : ty
-    -- --------------------------
-    --  E |- for x <- arr e : ty
-    doTypecheck for@(For {name, step, src, body}) =
-        do stepTyped <- doTypecheck step
-           srcTyped  <- doTypecheck src
-           let srcType = AST.getType srcTyped
-
-           unless (isArrayType srcType || isRangeType srcType) $
-             pushError src $ NonIterableError srcType
-
-           let elementType = if isRangeType srcType
-                             then intType
-                             else getResultType srcType
-           bodyTyped <- typecheckBody elementType body
-           return $ setType unitType for{step = stepTyped
-                                        ,src  = srcTyped
-                                        ,body = bodyTyped}
-        where
-          addIteratorVariable ty = extendEnvironmentImmutable [(name, ty)]
-          typecheckBody ty = local (addIteratorVariable ty) . typecheck
+
+    -- JOY for-comprehension
+    -- returnType is unittye, is changed during the second typechecking when it has been desugaraed into a method call
+    doTypecheck for@(For {sources, body}) = do
+      sourcesTyped <- typeCheckSources sources
+      forVarList <- getForVarTypeList sources
+      bodyTyped <- typecheckBody forVarList body
+      return $ setType unitType for{sources = sourcesTyped,
+                                      body = bodyTyped}
+      where
+        typeCheckSources :: [ForSource] -> TypecheckM [ForSource]
+        typeCheckSources sourceList =  do
+          typedSources <- mapM typeCheckSource sourceList
```

Comment:
if the return type is the result of the `mapM`, you move the line `mapM typCheckSource sourceList` to the line above, removing the `do` and the `return typeSources`

Reply:
If you do the comment above, maybe line 1729 is really a `mapM typeCheckSource sourceList`, and you have to maintain one less function